File: MonkeyGameJava/app/src/main/java/com/example/monkeygamejava/GameView.java
package com.example.monkeygamejava;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.view.SurfaceHolder;
import android.view.SurfaceView;
import android.view.View;
import android.widget.Toast;

import java.util.ArrayList;

import static java.security.AccessController.getContext;

public class GameView extends SurfaceView implements Runnable {
    public static int maxX = 20; // размер по горизонтали
    public static int maxY = 28; // размер по вертикали
    public static float unitW = 0; // пикселей в юните по горизонтали
    public static float unitH = 0; // пикселей в юните по вертикали
    private boolean firstTime = true;
    private boolean gameRunning = true;
    private Ship ship;
    private Thread gameThread = null;
    private Paint paint;
    private Canvas canvas;
    private SurfaceHolder surfaceHolder;

    private ArrayList<Asteroid> asteroids = new ArrayList<>(); // тут будут харанится астероиды
    private final int ASTEROID_INTERVAL = 50; // время через которое появляются астероиды (в итерациях)
    private int currentTime = 0;

    public GameView(Context context) {
        super(context);
        //инициализируем обьекты для рисования
        surfaceHolder = getHolder();
        paint = new Paint();

        // инициализируем поток
        gameThread = new Thread(this);
        gameThread.start();
    }
    private void checkCollision(){ // перебираем все астероиды и проверяем не касается ли один из них корабля
        for (Asteroid asteroid:asteroids) {
            if(asteroid.isCollision(ship.x, ship.y, ship.size) && asteroid.bitmapId!=R.drawable.fireball){
                GameActivity.setScore(5);
                asteroids.remove(asteroid);
                break;
            }
            else if(asteroid.isCollision(ship.x, ship.y, ship.size)&& asteroid.bitmapId==R.drawable.fireball){
                gameRunning=false;
                GameActivity.showScore();
            }
        }
    }

    private void checkIfNewAsteroid(){ // каждые 50 итераций добавляем новый астероид
        if(currentTime >= ASTEROID_INTERVAL){
            Asteroid asteroid = new Asteroid(getContext());
            asteroids.add(asteroid);
            currentTime = 0;
        }else{
            currentTime ++;
        }
    }
    public void run() {
        while (gameRunning) {
            update();
            draw();
            checkCollision();
            checkIfNewAsteroid();
            control();
        }
    }

    private void update() {
        if(!firstTime) {
            ship.update();
            for (Asteroid asteroid : asteroids) {
                asteroid.update();
            }
        }
    }

    private void draw() {
        if (surfaceHolder.getSurface().isValid()) {  //проверяем валидный ли surface

            if(firstTime){ // инициализация при первом запуске
                firstTime = false;
                unitW = surfaceHolder.getSurfaceFrame().width()/maxX; // вычисляем число пикселей в юните
                unitH = surfaceHolder.getSurfaceFrame().height()/maxY;
                ship = new Ship(getContext()); // добавляем корабль
            }

            canvas = surfaceHolder.lockCanvas(); // закрываем canvas
            canvas.drawColor(Color.TRANSPARENT);
            ship.drow(paint, canvas); // рисуем корабль

            for(Asteroid asteroid: asteroids){ // рисуем астероиды
                asteroid.drow(paint, canvas);
            }


            surfaceHolder.unlockCanvasAndPost(canvas); // открываем canvas
        }
    }

    private void control() { // пауза на 17 миллисекунд
        try {
            gameThread.sleep(15);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }
}

File: MonkeyGameJava/app/src/main/java/com/example/monkeygamejava/GameView.kt
package com.example.monkeygamejava

import android.content.Context
import android.graphics.*
import android.view.SurfaceHolder
import android.view.SurfaceView
import java.util.*

class GameView(context: Context?, private val callback: OnDieListener) : SurfaceView(context), Runnable {
    companion object{
        var maxX = 20
        var maxY = 28
        var unitW = 0f
        var unitH = 0f
    }

    private var firstTime = true
    private var gameRunning = true
    private var monkey: Monkey? = null
    private var gameThread: Thread? = null
    private val paint: Paint = Paint()
    private var canvas: Canvas? = null
    private val surfaceHolder: SurfaceHolder = holder
    private val items = ArrayList<FallItem>()
    private val ASTEROID_INTERVAL = 50
    private var currentTime = 0
    private fun checkCollision() {
        for (item in items) {
            if (item.isCollision(monkey!!.x, monkey!!.y, monkey!!.size)&&item.bitmapId == R.drawable.banan&& item.bitmapId != R.drawable.fireball && item.bitmapId != R.drawable.cold_fireball) {
                GameActivity.setScore(5)
                items.remove(item)
                break
            } else if (item.isCollision(monkey!!.x, monkey!!.y, monkey!!.size)&&item.bitmapId != R.drawable.banan && item.bitmapId != R.drawable.fireball && item.bitmapId != R.drawable.cold_fireball) {
                GameActivity.setScore(2)
                items.remove(item)
                break
            }else if (item.isCollision(monkey!!.x, monkey!!.y, monkey!!.size) && item.bitmapId == R.drawable.fireball) {
                callback.onDie()
                gameRunning = false
            }else if (item.isCollision(monkey!!.x, monkey!!.y, monkey!!.size) && item.bitmapId == R.drawable.cold_fireball) {
                callback.onDie()
                gameRunning = false
            }
        }
    }

    private fun checkIfNewAsteroid() {
        if (currentTime >= ASTEROID_INTERVAL) {
            val asteroid = FallItem(context)
            items.add(asteroid)
            currentTime = 0
        } else {
            currentTime++
        }
    }

    override fun run() {
        while (gameRunning) {
            update()
            draw()
            checkCollision()
            checkIfNewAsteroid()
            control()
        }
    }

    private fun update() {
        if (!firstTime) {
            monkey!!.update()
            for (asteroid in items) {
                asteroid.update()
            }
        }
    }

    private fun draw() {
        if (surfaceHolder.surface.isValid) {
            if (firstTime) {
                firstTime = false
                unitW = surfaceHolder.surfaceFrame.width() / maxX.toFloat()
                unitH = surfaceHolder.surfaceFrame.height() / maxY.toFloat()
                monkey = Monkey(context)
            }
            canvas = surfaceHolder.lockCanvas()
            canvas!!.drawColor(Color.TRANSPARENT, PorterDuff.Mode.CLEAR)
            monkey!!.drow(paint, canvas!!)
            for (asteroid in items) {
                asteroid.drow(paint, canvas!!)
            }
            surfaceHolder.unlockCanvasAndPost(canvas)
        }
    }

    private fun control() {
        try {
            Thread.sleep(15)
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
    }

    init {

        val sfvTrack: SurfaceView = this
        sfvTrack.setZOrderOnTop(true)
        val sfhTrackHolder = sfvTrack.holder
        sfhTrackHolder.setFormat(PixelFormat.TRANSPARENT)
        gameThread = Thread(this)
        gameThread!!.start()
    }
}

interface OnDieListener {
    fun onDie()
}